### DISCUSSION TITLE
How do I create a transparent png gradient file of 2000x3000 that looks like the attached image

### DISCUSSION THREAD

**Discussion by bullmoose20** [ImageMagick/ImageMagick#5688] [Help]
How do I create a transparent png gradient file of 2000x3000 that looks like the attached image. (Fade-Gradient.png)

I am still pretty new to Imagemagick and I am now using it to merge files together to create a final image, but I would like to be able to eliminate the need to have the Fade-Gradient amd produce it "on the fly"
Fade-Gradient:
![Fade-Gradient](https://user-images.githubusercontent.com/12549033/197207075-cd832d45-e3dd-4d51-8aa1-a2b4260e9c8e.png)
Base:
![base_#FB19B9](https://user-images.githubusercontent.com/12549033/197207630-cb643fec-bd04-4327-bfbf-d89e1302574f.jpg)
Base with Fade-Gradient:
![gradient_#FB19B9](https://user-images.githubusercontent.com/12549033/197207634-a26d4c4c-f038-49b9-826b-8c8e4f6a3b75.jpg)
Logo:
![resized_Spotify](https://user-images.githubusercontent.com/12549033/197207627-e2edab29-8f8e-4ffb-b84e-5b8b8ca0b920.png)
Final result with text and logo placement:
![Spotify](https://user-images.githubusercontent.com/12549033/197207632-fdbfaf8d-3008-438b-a163-088ceb7eaa4b.jpg)

Essentially, i use magick to merge the Fade-Gradient with the base to produce the gradient_base. Then I take the logo and add the text to come up with the final result. I want to create Fade-Gradient using Imagemagick.

**Comment by fmw42** [comment]
See https://imagemagick.org/Usage/canvas/#radial-gradient and https://imagemagick.org/script/gradient.php

**Reply by bullmoose20** [reply]
So i think that the radial gradient would replicate the middle of the Fade-Gradient.png file... and the last part is to apply the fade from northeast to southwest where the fade is slightly darker in the southwest corner compared to northeast.... Any tips on how I could accomplish the fade part?

**Reply by fmw42** [reply]
See the gradient link above for linear gradients and https://imagemagick.org/Usage/canvas/#gradient and https://imagemagick.org/Usage/canvas/#two_color_gradients

**Reply by fmw42** [reply]
Merge the linear and radial gradients to form your final result

**Reply by bullmoose20** [reply]
the struggle is real... :-)

![image](https://user-images.githubusercontent.com/12549033/197244400-9cbf7ce0-a535-4959-a2e0-c2535297b122.png)

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000  radial-gradient:none-blue              rgradient_range1.png

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000 gradient:none-black -fill firebrick -colorize 100% gradient_trans_colorize.png

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000 radial-gradient:none-black -fill firebrick -colorize 100% gradient_trans_colorize.png

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000 radial-gradient:none-black -fill white -colorize 100% gradient_trans_colorize.png

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000 radial-gradient:none-white -fill white -colorize 100% gradient_trans_colorize.png

PS C:\Users\zzz\Downloads\PMM-Posters\bullmoose20> magick -size 2000x3000 radial-gradient:none-white -fill black -colorize 100% gradient_trans_colorize.png


Just trying to get white in the middle and transparent on the outside....
The closest is the blue one, however the black is not transparent... so that does not work...

**Reply by bullmoose20** [reply]
And here is the best attempt at the fade from top right to bottom left with transparency

magick -size 2000x3000 gradient:none-black -rotate +45 -gravity center -crop 2000x3000+0+0 +repage gradient_diagonal.png
produces this:
![gradient_diagonal](https://user-images.githubusercontent.com/12549033/197247134-9fe9c021-ee7a-4e44-aaf1-21fdb873f824.png)

so it sort of works?

**Reply by bullmoose20** [reply]
magick -size 6500x6500 gradient:none-black -rotate +45 -gravity center -crop 2000x3000+0-2000 +repage gradient_diagonal.png
produced this... which I think is fine.. but this is quite slow to generate
![gradient_diagonal](https://user-images.githubusercontent.com/12549033/197250464-4e0821da-d21e-4d28-a427-336461fdc911.png)

**Reply by fmw42** [reply]
Use one of the diagonal gradient methods in the links provided.

`convert -define gradient:direction=southeast -size 500x500 gradient:black-white diag.png`

![diag](https://user-images.githubusercontent.com/13458218/197260734-f805d06e-23c9-4e0e-96c2-eb84fa137e17.png)


`convert -size 500x500 xc: -sparse-color Barycentric '0,0 black   499,499 white' diag2.png`

![diag2](https://user-images.githubusercontent.com/13458218/197260771-4d617dcd-72eb-4cdc-acfb-16ae981f41e3.png)

**Reply by bullmoose20** [reply]
I always need these to be transparent.... so I tried the first command:
` magick -define gradient:direction=southwest -size 2000x3000 gradient:none-black diag.png` and it gave me this (its too dark in the bottom left:
![diag](https://user-images.githubusercontent.com/12549033/197266726-ec9062da-5d6e-4f2f-9e6a-41173ac08ba5.png)

and for this one: `magick -size 2000x3000 xc: -sparse-color Barycentric '2000,0 white 0,6000 black' diag3.png` I got the following however, replacing black with none-black throws an error... so not sure how to make this one produce a transparent png:
![diag3](https://user-images.githubusercontent.com/12549033/197267201-55d1b489-8aa0-4588-a977-dcd91ad493f8.png)

**Comment by fmw42** [comment]
`convert -size 500x500 radial-gradient:white-black -alpha copy -depth 8 x.png`

![x](https://user-images.githubusercontent.com/13458218/197248188-4d8ef2c3-ca5d-4372-984c-0417bc282256.png)

Or the general way is to create a gradient alpha channel and then composite it over the background image. So if you want a red background fading to transparent

`convert -size 500x500 xc:red \( -size 500x500 radial-gradient:white-black \) -alpha off -compose copy_opacity -composite y.png
`
![y](https://user-images.githubusercontent.com/13458218/197249703-9fc5b0d5-047b-4d75-9f45-f262e66dae94.png)

**Reply by bullmoose20** [reply]
ok. let me try to use the second one and swap red for white

**Reply by fmw42** [reply]
How about this? Just make a gradient of the colors you desire and skip the alpha.

`convert -size 2000x3000 -define gradient:center=1000,1500 radial-gradient:"#FB19B9-#850E62" pink_grad.png`

![pink_grad](https://user-images.githubusercontent.com/13458218/197267668-a2bf4c1f-405d-41be-8d69-63937b8f5849.png)

**Reply by bullmoose20** [reply]
this is interesting.. how did you determine the two values for the gradient? "#FB19B9-#850E62"
The background will be a random color. So in the case here its that shade of pink, but in the script, it will be random...

**Reply by bullmoose20** [reply]
so for the script, the FB19B9 is what I randomly generate... its the other color #850E62 that I am wondering about...

Let me go see what that color is..

**Reply by bullmoose20** [reply]
looks like a darker shade of pink... so not sure how I would programmatically change the second value

**Reply by bullmoose20** [reply]
`magick -size 2000x3000 xc:white `( -size 2000x3000 radial-gradient:white-black `) -alpha off -compose copy_opacity -composite y.png` produces this which is way too bright:
![y](https://user-images.githubusercontent.com/12549033/197269498-dab05dda-95f7-4930-a5e3-0d6e58c02480.png)

**Reply by fmw42** [reply]
Is this what you want for the gray?

`convert -size 2000x3000 -define gradient:center=1000,1500 -define gradient:radii=1500,1500 radial-gradient:"gray(60%)-gray(40%)" gray_grad3.png`

![gray_grad3](https://user-images.githubusercontent.com/13458218/197273909-b962edb3-ec3d-4dfc-b1e4-dbbb6232a4eb.png)

**Reply by bullmoose20** [reply]
So the above is definitely less intense in terms of the radial-gradient, however, since there is no transparency, merging this with the base_solid pink would not work... and then finally to apply the 45 degree fade to dark would not work. Or am i missing something?

BTW, thanks so much for all your help and suggestions

**Comment by fmw42** [comment]
In the pink gradient, the center is the first color and corner is the second color. So change the hex values. You can use color names or srgb(red,green,blue) colors.  I simply measured your center and corner color with a color dropper tool off your image.

`convert -size 2000x3000 -define gradient:center=1000,1500 radial-gradient:"#FB19B9-#850E62" pink_grad.png`

**Reply by bullmoose20** [reply]
right.. so that won;t work as the value for the back.png is randomly generated and could be any hex color... so knowing what is in the corner does not help. this is why i was looking for a solution based on the transparent fade and radial gradient that you overlay on top of the base color.

**Reply by bullmoose20** [reply]
for the diagonal gradient, I think this is the best I can get:
![image](https://user-images.githubusercontent.com/12549033/197283794-2122e5c8-ec45-4854-87b0-58b25453b89e.png)

for the radial gradient in the middle... this is what happens when I try to merge the two layers:
![image](https://user-images.githubusercontent.com/12549033/197283995-7b6771a6-bb48-417d-9cca-7df9af158615.png)
or this:
![image](https://user-images.githubusercontent.com/12549033/197284077-4e6cb118-4e19-44c4-90c2-c598a7be51fd.png)

**Reply by bullmoose20** [reply]
and here is what Fade-Gradient.png looks like in the same editor:
![image](https://user-images.githubusercontent.com/12549033/197284307-d4ba1ac8-5761-448b-a561-a636d0aeed80.png)

**Reply by bullmoose20** [reply]
ever so slightly bright in the middle with diagonal gradient top-right(brightest) to bottom left (darkest)

**Comment by fmw42** [comment]
Does this do what you want?  This is transparent.

`convert -size 2000x3000 xc:black \( -size 2000x3000 radial-gradient:"gray(10%)-gray(45%)" \) -alpha off -compose copy_opacity -composite gray_rad_grad.png`

![gray_rad_grad](https://user-images.githubusercontent.com/13458218/197290669-f3b0a4b7-ad4c-4152-81ea-10bdc6c2fc81.png)

I leave it to you to tweak the two gray percentages as desired.

**Reply by bullmoose20** [reply]
this is pretty darn close... i think the only issue is from what I see, the darkening in each corner is equal.... whereas in the original, there is a fade from top right to bottom left...so maybe I should try to combine this with the other diagonal gradient to see what it looks like

**Comment by snibgo** [comment]
> How do I create a transparent png gradient file of 2000x3000 that looks like the attached image. (Fade-Gradient.png)

Where did that image come from? It is black, with alpha that varies between about 10.6% and 65.9%. That variation is like a radial gradient, with the lowest value at the center, extending to the corners. However, the gradient (ie the rate of change) varies at different radii (ie in different directions), so the values in the four corners are different. In addition, it is noisy. In addition, there is a strange border effect.

How much of this do you want in a "looks like" result? "`-define gradient:extent=Diagonal radial-gradient:Black-White`" etc can simulate the main effect. The others are more complex.

**Reply by bullmoose20** [reply]
someone produced it for me a while ago in photoshop. I might be able to ask them to give me the psd if that would help?

**Comment by snibgo** [comment]
The PSD wouldn't tell me if those features (different values at corners, noise and border) are important to you.

**Comment by fmw42** [comment]
Why does it have to be a precise match for what you have from Photoshop?  What is close enough?

**Reply by fmw42** [reply]
Post your PSD file for the radial gradient and I will take a look at how it was created.

**Comment by bullmoose20** [comment]
I think that if the radial gradient you  shared with the variable gray percentages had a diagonal gradient applied northeast lighter than southwest, it would be fine. Two transparent layers that can then be merged into a final file that looks similar to the Fade-Gradient.png would be great.

The reason it needs to be similar is that we are going to apply this to a subset of images (3000 out of 6000) and wanted to ensure that the final results will be similar so that the effect looks the same for all 6000 images.

________________________________

**Reply by fmw42** [reply]
If you use the same Imagemagick filter all the time, they will be similar whether there is a linear gradient or not.  My question was why does it have to match the photoshop filter so close?

I will try to add that tomorrow sometime. But post your PSD file for how you created the gradient image.

**Comment by bullmoose20** [comment]
You are amazing for all the help you have been giving. Thanks so much. I thought I had answered the “why does it need to be so close?”

I can explain…
At first, I was manually applying the radial gradient filter in gimp because I did not notice the linear diagonal gradient. However, the author of the images noticed it and asked to adjust. So in gimp, I added the additional layer to get the result.  I then proceeded with a solution that was ok but it’s quite a long process in gimp to achieve the results that the author wanted. He provided the image I shared here and have been using that with imagemagick and a powershell wrapper script that will create the final results. The issue is that the png with the gradients is always misplaced or stored in the wrong location when others use the wrapper. By creating it on the fly, that issue goes away and anyone that uses the wrapper powershell script will create the final images as the author intended and respecting the theme of the images. There are thousands of them and we also need to recreate them as many are incorrect…about 3000 out of 6000.
________________________________

**Comment by snibgo** [comment]
We can extract the alpha channel from Fade-Gradient.png:
```
magick Fade-Gradient.png -alpha extract f0.png
```
f0.png is a grayscale image. The task is to reproduce it with IM.

As I have said, it is like a radial gradient, with the lowest value at the center, extending to the corners. However, the gradient (ie the rate of change) varies at different radii (ie in different directions). In addition, it is noisy. In addition, there is a strange border effect.

If we look at only the radial gradient, the average of the corners is roughly 53.2127%, and the average of the centre is roughly 10.8902%. So a gradient image is (Windows BAT syntax):
```
magick ^
  -size 2000x3000 ^
  -define gradient:extent=Diagonal ^
  radial-gradient:gray(10.8902%%)-gray(53.2127%%) ^
  s2.png
```
The RMSE from f0.png is 7.6%. The centre is accurate, but the four corners are equal when they shouldn't be. We can correct the corners by multiplying a radial gradient by a bilinear gradient:
```
magick ^
  -size 2000x3000 ^
  ( -define gradient:extent=Diagonal ^
    radial-gradient:gray(10.8902%%)-gray(64.3020%%) ^
  ) ^
  ( xc: ^
    -sparse-color bilinear "0,0,gray(73.6355%%), %%[fx:w-1],0,gray(70.0798%%), 0,%%[fx:h-1],gray(100%%) %%[fx:w-1],%%[fx:h-1],gray(87.3026%%)" ^
  ) ^
  -compose Multiply -composite ^
  s3.png
```
The RMSE from f0.png is 6.5%. The centre and four corners are accurate. This result may be good enough.

**Reply by bullmoose20** [reply]
hi there... I wanted to share what I did just to give you an idea... and of the results... I have not tried what you did above, however, this is what I found.

`magick -size 6500x6500 gradient:none-black -rotate +45 -gravity center -crop 2000x3000+0-2000 +repage gradient_diagonal.png`
`magick -size 2000x3000 xc:black `( -size 2000x3000 radial-gradient:"gray(10%)-gray(45%)" `) -alpha off -compose copy_opacity -composite gray_rad_grad.png`
`magick .\gradient_diagonal.png .\gray_rad_grad.png -background None -layers Flatten .\Fade-Gradient2.png`

Then I ran the script to add the base color, logo, and text and this is the result. (v2) are the files used with the Fade-Gradient2.png and the ones without v2 are based on the premade Fade-Gradient.png

v1
![Apple](https://user-images.githubusercontent.com/12549033/197396391-306d4c1e-f52d-4a66-94c4-b41a29dfce8b.jpg)
v2
![Applev2](https://user-images.githubusercontent.com/12549033/197396392-8731f8d9-2783-4ce2-a30b-9560f35a8d2e.jpg)
v1
![Spotify](https://user-images.githubusercontent.com/12549033/197396393-2c5053b6-6652-4480-81db-bab4e79d06ee.jpg)
v2
![Spotifyv2](https://user-images.githubusercontent.com/12549033/197396394-3e7fc9f8-3ec6-4914-9dc3-73fb3c13910f.jpg)

As you can see, the v2 are a shade darker, but they look pretty close to the original IMHO.

**Comment by snibgo** [comment]
> ... but they look pretty close to the original IMHO.

Are they good enough? If they are, then the problem is solved.

**Reply by bullmoose20** [reply]
Any tips on optimizing the commands? The three in unison take a  total of 20 seconds to run.

```
PS C:\Users\bullmoose20\Downloads\PMM-Posters\bullmoose20> measure-command { magick -size 6300x6300 gradient:none-black -rotate +45 -gravity center -crop 2000x3000+0-2000 +repage gradient_diagonal.png }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 14
Milliseconds      : 392
Ticks             : 143920601
TotalDays         : 0.000166574769675926
TotalHours        : 0.00399779447222222
TotalMinutes      : 0.239867668333333
TotalSeconds      : 14.3920601
TotalMilliseconds : 14392.0601




PS C:\Users\bullmoose20\Downloads\PMM-Posters\bullmoose20> measure-command { magick -size 2000x3000 xc:black `( -size 2000x3000 radial-gradient:"gray(10%)-gray(45%)" `) -alpha off -compose copy_opacity -composite gray_rad_grad.png }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 3
Milliseconds      : 608
Ticks             : 36084342
TotalDays         : 4.17642847222222E-05
TotalHours        : 0.00100234283333333
TotalMinutes      : 0.06014057
TotalSeconds      : 3.6084342
TotalMilliseconds : 3608.4342




PS C:\Users\bullmoose20\Downloads\PMM-Posters\bullmoose20> measure-command { magick .\gradient_diagonal.png .\gray_rad_grad.png -background None -layers Flatten .\Fade-Gradient2.png }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 1
Milliseconds      : 786
Ticks             : 17865263
TotalDays         : 2.06773877314815E-05
TotalHours        : 0.000496257305555556
TotalMinutes      : 0.0297754383333333
TotalSeconds      : 1.7865263
TotalMilliseconds : 1786.5263
```

**Comment by fmw42** [comment]
You have two issues with timing.  First, you need to combine the 3 commands into one to avoid writing temporary images to disk.  Second you are wasting a lot of time creating a 6500x6500 image only to crop the center 2000x3000. So I saved your cropped image and measured the colors in the top left and bottom right corners and just made a 2000x3000 gradient with those colors in the corner directly.

This is I believe the Windows equivalent of my bash command:

```
magick ( -size 2000x3000 gradient:"graya(0,0.227)-graya(0,0.337)" ) ^
( -size 2000x3000 xc:black ( -size 2000x3000 radial-gradient:"gray(10%)-gray(45%)" ) -alpha off -compose copy_opacity -composite ) ^
-background None -layers Flatten Fade-Gradient2.png
```

![Fade-Gradient2](https://user-images.githubusercontent.com/13458218/197404063-dae56aaf-7eb0-4767-8bbc-7a3533c224ea.png)

On my Mac, my command took about 3 seconds.

Is this now close to what you want? If not, then edit any of the 4 arguments for the 2 gradients to make it closer.

**Comment by bullmoose20** [comment]
Hi there... How did you measure the corners to come up with ` -size 2000x3000 gradient:"graya(0,0.227)-graya(0,0.337)" `?
I ask because the gradient should be from top right to bottom left... and I think this is the problem with the above... Maybe color picker would help? Not sure how it translates to 0.227 and 0.337

**Reply by bullmoose20** [reply]
So I have been messing around with the above and the values... this is not a diagonal gradient from top right to bottom left. This is a North to South gradient. The reason I had it at 45 degrees was to make it lighter in the top right and then darker in the bottom left. Let me see if changing the coordinates will do anything

**Reply by fmw42** [reply]
Sorry, I measured the wrong corners. I thought you wanted top-left and bottom-right. I also left off the gradient define to set the direction for the values I got. ( I note that there is a bug in that the linear gradient direction is affecting the radial gradient. So I had to reset the center with another define).

Try this

```
magick ( -size 2000x3000 -define gradient:direction=southwest gradient:"graya(0,0.0117647)-graya(0,0.552941)" ) ^
( -size 2000x3000 xc:black ( -size 2000x3000 -define gradient:center=1000x1500 radial-gradient:"gray(10%)-gray(45%)" ) -alpha off -compose copy_opacity -composite ) ^
-background None -layers Flatten Fade-Gradient2.png

```

![Fade-Gradient2](https://user-images.githubusercontent.com/13458218/197405917-493a53e3-1420-4a6c-bac6-5ef3f2a44608.png)


Alternately, to avoid the linear gradient direction from affecting the radial gradient, we can add -respect-perenthesis to the command line. So this should give the same result:

```
magick -respect-parenthesis ( -size 2000x3000 -define gradient:direction=southwest gradient:"graya(0,0.0117647)-graya(0,0.552941)" ) ^
( -size 2000x3000 xc:black ( -size 2000x3000 radial-gradient:"gray(10%)-gray(45%)" ) -alpha off -compose copy_opacity -composite ) ^
-background None -layers Flatten Fade-Gradient2.png

```

**Reply by bullmoose20** [reply]
In the end, I went with the following:
```
  magick -size 6500x6500 gradient:none-black -rotate +45 -gravity center -crop 2000x3000+0-2000 +repage gradient_diagonal.png
  magick gradient_diagonal.png -level 0%,100%,0.5 gradient_diagonal.png
  magick -size 2000x3000 xc:black `( -size 2000x3000 radial-gradient:"gray(5%)-gray(45%)" `) -alpha off -compose copy_opacity -composite gray_rad_grad.png
  magick gradient_diagonal.png gray_rad_grad.png -background None -layers Flatten Fade-Gradient.png
```
These 4 commands produce a close replica of the original Fade-Gradient posted above. Then the script will just reference this and add logo and text and border without the expensive operation of running all 4 of these commands each time. First time it takes about 20 seconds and after that, less than 3 to process.
Thanks to both of you for all your help!

**Reply by fmw42** [reply]
You are wasting time by writing each to disk and then reading them back again.  You can make this all one long command line.
